### DISCUSSION TITLE
Example constraint to restrict workspaces from depending on different version of a package

### DISCUSSION THREAD

**Discussion by wdfinch** [yarnpkg/berry#2693] [Q&A] [answered]
I would like to write a constraint to enforce that a specific version of a package is used (if it's used at all). So for example, if a workspace includes `react` as a dependency, then I would like to enforce that only `react` 17.0.2 may be used.

While I understand how this works in theory and the docs are helpful, I'm struggling to write this constraint as I'm not sure on the exact syntax to use in a real-world example.

I did see this recipe which looks like what I want to do, but I'm not sure if this is the right choice. I would really appreciate it if someone could write the above constraint out for me so I could see a real-world example of this. 

```
gen_enforced_dependency(WorkspaceCwd, DependencyIdent, DependencyRange2, DependencyType) :-
  workspace_has_dependency(WorkspaceCwd, DependencyIdent, DependencyRange, DependencyType),
  workspace_has_dependency(OtherWorkspaceCwd, DependencyIdent, DependencyRange2, DependencyType2),
  DependencyRange \= DependencyRange2.
```

**Comment by arcanis** [comment, marked as answer]
Didn't try it, but it should be something like this:

```prolog
gen_enforced_dependency(WorkspaceCwd, 'react', '17.0.2', DependencyType) :-
  workspace_has_dependency(WorkspaceCwd, 'react', _, DependencyType).
```

**Reply by wdfinch** [reply]
Thanks!

**Reply by wdfinch** [reply]
Aside from the recipes, it would be cool if the docs included a bunch of examples of real-world constraints. Most people using yarn 2 will not know prolog and it's not easy to figure out how it works. This could help us get started.
